static/js/ace-editor.ts

```diff
@@ -233,91 +188,44 @@ export class HedyAceEditor implements HedyEditor {
   clearErrors(): void {
     // Not sure if we use annotations everywhere, but this was
     // here already.
-    this._editor?.session.clearAnnotations();
+    this._editor.session.clearAnnotations();
     this.markers?.clearErrors();
   }
 
   /**     
    * Moves to the cursor to the end of the current file
    */
   moveCursorToEndOfFile(): void { 
-    this._editor?.navigateFileEnd();
+    this._editor.navigateFileEnd();
   }
 
   /**
    * Clears the selected text
    */
   clearSelection(): void {
-    this._editor?.clearSelection();
+    this._editor.clearSelection();
   }
 
   /**
   * Removes all breakpoints on the rows.
   **/
   clearBreakpoints(): void { 
-    this._editor?.session.clearBreakpoints();
+    this._editor.session.clearBreakpoints();
   }
 
   /**
    * If this editor is used as a main editor, we set the options here
    */
   configureMainEditor(): void {
-    this._editor?.setShowPrintMargin(false);
-    this._editor?.renderer.setScrollMargin(0, 0, 0, 20)
-    this._editor?.addEventListener('change', () => {
-      theLocalSaveWarning.setProgramLength(this._editor!.getValue().split('\n').length);
-    });
-    // Set const value to determine the current page direction -> useful for ace editor settings
-    const dir = $("body").attr("dir");
-    if (dir === "rtl") {
-      this._editor?.setOptions({ rtl: true });
-    }
-
-    // If prompt is shown and user enters text in the editor, hide the prompt.
-    this._editor?.on('change', () => {
-      if (this.askPromptOpen) {
-        stopit();
-        this._editor?.focus(); // Make sure the editor has focus, so we can continue typing
-      }
-      if ($('#ask-modal').is(':visible')) $('#inline-modal').hide();
-      this.askPromptOpen = false;
-      $('#runit').css('background-color', '');
-      this.clearErrors();
-      //removing the debugging state when loading in the editor
-      stopDebug();
+    this._markers = new Markers(this._editor);
+    // *** Debugger *** //
+    // TODO: FIX THIS
+    initializeDebugger({
+      editor: this._editor,
+      markers: this.markers,
+      level: theLevel,
+      language: theLanguage,
```

Comment:
Yeah, it's interesting to think about whether the debugger should be inside the editor, or outside it.

You *could* say that the debugger is external to the editor, just using some features of the editor like "putting markers in the gutter" and "striking through certain lines".

But no need to change that now. You can always do that later, once you start adding a new editor implementation and you notice that there's a lot of repeated code between them 😉

Reply:
Yes, exactly, this is a weird one. Because yeah, I can say that the debugger is part of an editor so it should be part of this interface. On the other hand, it'll probably have to deal with HTML and DOM objects, and we're trying to avoid to include those things in this interface. 

However, I think that will become clearer once I start working on including code mirror and replacing the current debugger.